---
title: "Seurat Brief Guide"
author: "Celine Garcia"
date: "2025-11-05"
output: html_document
---

Here is a brief Introduction to Seurat. This document shows the basic functions and workflow that can be use to analyse single-cell data. [PBMC 3K Guided tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial), [Using sctransform in Seurat](https://satijalab.org/seurat/articles/sctransform_vignette) and [Essential Commands](https://satijalab.org/seurat/articles/essential_commands) vignettes had been used. Raw data can be found [here.](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz)

# Install Seurat

```{r}
install.packages('Seurat') # It requieres around 30-45 min. 
```

For enhance speed and performance, install:

```{r}
setRepositories(ind = 1:3, addURLs = c('https://satijalab.r-universe.dev', 'https://bnprks.r-universe.dev/'))
install.packages(c("BPCells", "presto", "glmGamPoi"))

# Install the remotes package
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
install.packages('Signac')
remotes::install_github("satijalab/seurat-data", quiet = TRUE)
remotes::install_github("satijalab/azimuth", quiet = TRUE)
remotes::install_github("satijalab/seurat-wrappers", quiet = TRUE)
```

## Package

For this session, use the following libraries:

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(SeuratData)
```

## Seurat Object

```{r}
# Si esta comprimido
untar("home/biouser/Desktop/Celine/SEURAT/pbmc3k_filtered_gene_bc_matrices.tar.gz", exdir =".")

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "/home/biouser/Desktop/Celine/SEURAT/filtered_gene_bc_matrices/hg19/")

# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
```

This creates an object of class Seurat. In this case, we have 1 layer of counts present.

The recent version of CellRanger now also output using the h5_file_format which can be read in using `Read10X_h5`

For examine the data of the first 30 cells:

```{r}
pbmc.data[c("CD3D", "TCL1A", "MS4A1"), 1:30]
```

The . values represent 0s (no molecules detected). Since most values in a snRNA-seq matrix are 0, Seurat uses a sparse matrix representation whenever is possible. This results in significant memory and speed savings for Drop-seq/inDrop/10xdata.

```{r}
dense.size <- object.size(as.matrix(pbmc.data)) # Inneficient dense matrix. Keeps all values
sparse.size <- object.size(pbmc.data) # Efficient matrix: keeps the positions and values higher then 0.
dense.size/sparse.size # shows the relation between the dense and sparse matrix.
```

# Pre-processing Workflow

## QC and selecting cells

QC metrics include:

-   The number of unique genes detected in each cell.

    -   Low-quality cells or empty droplets will often have very few genes

    -   Cell doublets or multiplets may exhibit an aberrantly high gene count

-   Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)

-   The percentages of reads that map to the mitochondrial genome

    -   Low-quality / dying cells often exhibit extensive mitochondrial contamination

    -   We calculate mitochondrial QC metrics with the [`PercentageFeatureSet()`](https://satijalab.org/seurat/reference/percentagefeatureset) function, which calculates the percentage of counts originating from a set of features

    -   We use the set of all genes starting with `MT-` as a set of mitochondrial genes

```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")
```

-   The number of unique genes and total molecules are automatically calculated during [`CreateSeuratObject()`](https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html)

    -   You can find them stored in the object meta data

```{r}
head(pbmc@meta.data, 5)
```

Some plots:

```{r}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

In the example below, we visualize QC metrics, and use these to filter cells.

-   We filter cells that have unique feature counts over 2,500 or less than 200

-   We filter cells that have \>5% mitochondrial counts

```{r}
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

## **Standard Workflow**

### Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “*LogNormalize*” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. In Seurat v5, Normalized values are stored in `pbmc[["RNA"]]$data`.

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000)

# By default
pbmc <- NormalizeData(pbmc)
```

While this method of normalization is standard and widely used in scRNA-seq analysis, global-scaling relies on an assumption that each cell originally contains the same number of RNA molecules.

They have developed alternative workflows for the single cell preprocessing that do not make these assumptions. The use of `SCTransform` replaces the need to run `NormalizeData`, `FindVariableFeatures`, or `ScaleData` (described below.)

### Identification of highly variable features (feature selection)

This calculates the subset of features that exhibit high cell-to-cell variation in the dataset. Directly modeling the mean-variance relationship inherent in single-cell data and is implemented in the `FindVariableFeatures()` function. By default, they return 2000 features per dataset.

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE, force = 5)
plot1 + plot2
```

### Scaling the data

Scaling is referring to linear transformation that is standard pre-processing step prior to dimensional reduction techniques like PCA. The [`ScaleData()`](https://satijalab.org/seurat/reference/scaledata) function:

-   Shifts the expression of each gene, so that the mean expression across cells is 0

-   Scales the expression of each gene, so that the variance across cells is 1

    -   This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate

-   The results of this are stored in `pbmc[["RNA"]]$scale.data`

-   By default, only variable features are scaled.

-   You can specify the `features` argument to scale additional features

```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes)
```

For remove unwanted sources of variation as heterogeneity associated with cell cycle stage or mitochondrial contamination:

```{r}
pbmc <- ScaleData(pbmc, vars.to.regress = "percent.mt")
# Diferentes Features: La advertencia te indica que el conjunto de genes (features) que se está escalando ahora es diferente del conjunto de genes que fue escalado en la ejecución anterior de ScaleData
```

# Linear dimensional reduction (PCA)

As input we have the previous determined variable features. If you want to use another different, it must pass `ScaleData` first.

For the first principal components, Seurat outputs a list of genes with the most positive and negative loadings, representing modules of genes that exhibit either correlation (or anti-correlation) across single-cells in the dataset.

```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc))
```

Seurat provides several useful ways of visualizing both cells and features that define the PCA, including `VizDimReduction()`, [`DimPlot()`](https://satijalab.org/seurat/reference/dimplot), and [`DimHeatmap()`](https://satijalab.org/seurat/reference/dimheatmap)

```{r}
# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5)
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
DimPlot(pbmc, reduction = "pca") + NoLegend()
```

[`DimHeatmap()`](https://satijalab.org/seurat/reference/dimheatmap) allows an easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses.

Both cells and features are ordered according to their PCA scores.

```{r}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE)
```

```{r}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

### Dimensionality

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores. The top principal components therefore represent a robust compression of the dataset. However, how many components should we choose to include? 10? 20? 100?

They used to apply `JackStraw` procedure but, because of the slowness and computationally expensive procedure, an alternative heuristic method generates an Elbowplot: a ranking of principle components based on the percentage of variance explained by each one ([`ElbowPlot()`](https://satijalab.org/seurat/reference/elbowplot) function).

In this example, we can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the **first 10 PCs.**

```{r}
ElbowPlot(pbmc)
```

Identifying the true dimensionality of a dataset can be challenging/uncertain. We therefore suggest these multiple approaches for users.

1.  PCs determine relevant sources of heterogeneity and could be used in conjunction with GSEA for example.
2.  `ElbowPlot`()
3.  The use of a common heuristic that can be calculated instantly. In this example, we might have been justified in choosing anything between PC 7-12 as a cutoff.

We chose 10 here, but encourage users to consider the following:

-   Dendritic cell and NK aficionados may recognize that genes strongly associated with PCs 12 and 13 define rare immune subsets. However, these groups are so rare, they are difficult to distinguish from background noise for a dataset of this size without prior knowledge.

-   Repeat downstream analyses with a different number of PCs (10, 15, or even 50!).

-   Err on the higher side when choosing this parameter. For example, performing downstream analyses with only 5 PCs does significantly and adversely affect results.

# Cluster the cells

Seurat applies a graph-based clustering approach, building upon initial strategies in ([Macosko *et al*](http://www.cell.com/abstract/S0092-8674(15)00549-8)). Importantly, the *distance metric* which drives the clustering analysis (based on previously identified PCs) remains the same.

Our newest approach embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’.

1.  Construct a KNN graph based on the euclidean distance in PCA space and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods –\> `FindNeighbors()` :takes as input the previously defined dimensionality of the dataset (first 10 PCs).
2.  To cluster the cells, we next apply modularity optimization techniques such as the Louvain algorithm (default) or SLM, to iteratively group cells together, with the goal of optimizing the standard modularity function. –\> [`FindClusters()`](https://satijalab.org/seurat/reference/findclusters) : contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. Between **0.4-1.2** typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets.
3.  The clusters can be found using the [`Idents()`](https://satijalab.github.io/seurat-object/reference/Idents.html) function.

```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:10)
pbmc <- FindClusters(pbmc, resolution = 0.5)
head(Idents(pbmc), 5)
```

## Non-linear dimensional reduction (UMAP/tSNE)

Seurat offers several non-linear dimensional reduction techniques, such as tSNE and UMAP. The goal of these algorithms is to learn underlying structure in the dataset, in order to place similar cells together in low-dimensional space. Therefore, cells that are grouped together within graph-based clusters determined above should co-localize on these dimension reduction plots.

These methods aim to preserve local distances in the dataset (i.e. ensuring that cells with very similar gene expression profiles co-localize), but often do not preserve more global relationships. We encourage users to leverage techniques like UMAP for visualization, but to avoid drawing biological conclusions solely on the basis of visualization techniques.

```{r}
pbmc <- RunUMAP(pbmc, dims = 1:10)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")
```

For saving:

```{r}
saveRDS(pbmc, file = "../output/pbmc_tutorial.rds")
```

# Finding differentially expressed features (cluster biomarkers)

## Find All Markers

Seurat can help you find markers that define clusters via differential expression (DE). By default, it identifies positive and negative markers of a single cluster (specified in `ident.1`), compared to all other cells. [`FindAllMarkers()`](https://satijalab.org/seurat/reference/findallmarkers) automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

```{r}
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2)
head(cluster2.markers, n = 5)

# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

Seurat has several tests for differential expression which can be set with the test.use parameter (see our [DE vignette](https://satijalab.org/seurat/articles/de_vignette) for details). For example, the ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

We include several tools for visualizing marker expression:

-   `VlnPlot()`--\> shows expression probability distributions across clusters

-   [`FeaturePlot()`](https://satijalab.org/seurat/reference/featureplot) –\> visualizes feature expression on a tSNE or PCA plots

-   There are other types of plots as [`RidgePlot()`](https://satijalab.org/seurat/reference/ridgeplot), [`CellScatter()`](https://satijalab.org/seurat/reference/cellscatter), and [`DotPlot()`](https://satijalab.org/seurat/reference/dotplot)

-   [`DoHeatmap()`](https://satijalab.org/seurat/reference/doheatmap) generates an expression heatmap for given cells and features.

```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A"))
```

Raw data plots:

```{r}
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```

```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", "CD8A"))
```

```{r}
#In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend()
```

## Presto

In Seurat v5, they use the presto package because it improves the speed of DE analysis, particularly for large datasets. **Only if you have cell_type column.**

```{r}
devtools::install_github("immunogenomics/presto")
```

```{r}
library(presto)
#Checkin if it is a Seurat objetc
object_seurat <- Seurat::UpdateSeuratObject(pbmc)
```

`seurat_assay` specified the assay used by `wilcoxauc.`

```{r}
head(wilcoxauc(object_seurat, 'cell_type', seurat_assay = 'RNA'))

```

For other features as counts, data (raw data) or scaled data, use `assay = 'counts' // assay = 'data' // assay = 'scale.data'`

# SCTransform

Biological heterogeneity in single-cell RNA-seq data is often confounded by technical factors including sequencing depth. The number of molecules detected in each cell can vary significantly between cells, even within the same celltype. Interpretation of scRNA-seq data requires effective pre-processing and normalization to remove this technical variability.

In [our manuscript](https://genomebiology-biomedcentral-com/articles/10.1186/s13059-021-02584-9) we introduce a modeling framework for the normalization and variance stabilization of molecular count data from scRNA-seq experiments. This procedure omits the need for heuristic steps including pseudocount addition or log-transformation and improves common downstream analytical tasks such as variable gene selection, dimensional reduction, and differential expression. We named this method `sctransform`.

```{r}
library(ggplot2)
library(sctransform)
```

### Apply sctransform normalization

This single command replaces [`NormalizeData()`](https://satijalab.org/seurat/reference/normalizedata), [`ScaleData()`](https://satijalab.org/seurat/reference/scaledata), and [`FindVariableFeatures()`](https://satijalab.org/seurat/reference/findvariablefeatures).Transformed data will be available in the SCT assay, which is set as the default after running sctransform.

During normalization, we can also remove confounding sources of variation, for example, mitochondrial mapping percentage

```{r}
# store mitochondrial percentage in object meta data
pbmc <- PercentageFeatureSet(pbmc, pattern = "^MT-", col.name = "percent.mt")

# run sctransform
pbmc <- SCTransform(pbmc, vars.to.regress = "percent.mt", verbose = FALSE)
```

### Perform dimensionality reduction by PCA and UMAP embedding

```{r}
# These are now standard steps in the Seurat workflow for visualization and clustering
pbmc <- RunPCA(pbmc, verbose = FALSE)
pbmc <- RunUMAP(pbmc, dims = 1:30, verbose = FALSE)

pbmc <- FindNeighbors(pbmc, dims = 1:30, verbose = FALSE)
pbmc <- FindClusters(pbmc, verbose = FALSE)
DimPlot(pbmc, label = TRUE)
```

In the [Standard Workflow], we use 10 PCs for this dataset, though results remain consistent with higher values. When using sctransform, increasing this number often improves results because it provides stronger normalization and removes more technical noise.

Even after standard log-normalization, sequencing depth can still affect higher PCs, while sctransform largely eliminates this issue. As a result, higher PCs more accurately capture subtle biological variation, improving downstream analyses.

Additionally, sctransform selects 3,000 variable features by default (vs. 2,000), since the extra features better reflect biological rather than technical differences. Overall, sctransform results are more stable across parameter choices and can even use all genes with minimal impact on outcome—though at a computational cost. This robustness allows users to apply consistent workflows efficiently across new datasets.

For example, the following code replicates the full end-to-end workflow, in a single command:

```{r}
pbmc <- CreateSeuratObject(pbmc_data) %>%
    PercentageFeatureSet(pattern = "^MT-", col.name = "percent.mt") %>%
    SCTransform(vars.to.regress = "percent.mt") %>%
    RunPCA() %>%
    FindNeighbors(dims = 1:30) %>%
    RunUMAP(dims = 1:30) %>%
    FindClusters()
```

The normalized values are stores in the SCT\_ assay.

-   `pbmc[["SCT"]]$scale.data` contains the residuals (normalized values), and is used directly as input to PCA. Please note that this matrix is non-sparse, and can therefore take up a lot of memory if stored for all genes. To save memory, we store these values only for variable genes, by setting the return.only.var.genes = TRUE by default in the [`SCTransform()`](https://satijalab.org/seurat/reference/sctransform) function call.

-   To assist with visualization and interpretation, we also convert Pearson residuals back to ‘corrected’ UMI counts. You can interpret these as the UMI counts we would expect to observe if all cells were sequenced to the same depth. If you want to see exactly how we do this, please look at the correct function [here](https://github.com/ChristophH/sctransform/blob/master/R/denoise.R).

-   The ‘corrected’ UMI counts are stored in `pbmc[["SCT"]]$counts`. We store log-normalized versions of these corrected counts in `pbmc[["SCT"]]$data`, which are very helpful for visualization.

-   sctransform normalization reveals sharper biological distinctions

```{r}
# These are now standard steps in the Seurat workflow for visualization and clustering
# Visualize canonical marker genes as violin plots.
VlnPlot(pbmc, features = c("CD8A", "GZMK", "CCL5", "S100A4", "ANXA1", "CCR7", "ISG15", "CD3D"),
    pt.size = 0.2, ncol = 4)
```

# Assigning cell type identity to clusters

Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:

| Cluster ID | Markers       | Cell Type    |
|:-----------|:--------------|:-------------|
| 0          | IL7R, CCR7    | Naive CD4+ T |
| 1          | CD14, LYZ     | CD14+ Mono   |
| 2          | IL7R, S100A4  | Memory CD4+  |
| 3          | MS4A1         | B            |
| 4          | CD8A          | CD8+ T       |
| 5          | FCGR3A, MS4A7 | FCGR3A+ Mono |
| 6          | GNLY, NKG7    | NK           |
| 7          | FCER1A, CST3  | DC           |
| 8          | PPBP          | Platelet     |

```{r}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

```{r}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + guides(colour = guide_legend(override.aes = list(size = 10)))
plot
#ggsave(filename = "../output/images/pbmc3k_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)
```

# Visualization

```{r}
#Ejemplo
SeuratData::InstallData("pbmc3k")
pbmc3k.final <- LoadData("pbmc3k", type = "pbmc3k.final")
pbmc3k.final$groups <- sample(c("group1", "group2"), size = ncol(pbmc3k.final), replace = TRUE)
features <- c("LYZ", "CCL5", "IL32", "PTPRCAP", "FCGR3A", "PF4")
```

1.  Ridge plots

    ```{r}
    # Ridge plots - from ggridges. Visualize single cell expression distributions in each cluster
    RidgePlot(pbmc3k.final, features = features, ncol = 2)
    ```

2.  Violinplots

    ```{r}
    # Violin plot - Visualize single cell expression distributions in each cluster
    VlnPlot(pbmc3k.final, features = features)
    ```

    ```{r}
    # Violin plots can also be split on some variable. Simply add the splitting variable to object
    # metadata and pass it to the split.by argument
    VlnPlot(pbmc3k.final, features = "percent.mt", split.by = "groups")
    ```

3.  Feature plot

    ```{r}
    # Feature plot - visualize feature expression in low-dimensional space
    FeaturePlot(pbmc3k.final, features = features)
    ```

    ```{r}
    # Plot a legend to map colors to expression levels
    # Adjust the contrast in the plot (min/max.cutoff)
    FeaturePlot(pbmc3k.final, features = "MS4A1", min.cutoff = 1, max.cutoff = 3)
    ```

    ```{r}
    # Visualize co-expression of two features simultaneously
    FeaturePlot(pbmc3k.final, features = c("MS4A1", "CD79A"), blend = TRUE)
    ```

    ```{r}
    # Split visualization to view expression by groups (replaces FeatureHeatmap)
    FeaturePlot(pbmc3k.final, features = c("MS4A1", "CD79A"), split.by = "groups")
    ```

4.  Dotplot

    ```{r}
    # Dot plots - the size of the dot corresponds to the percentage of cells expressing the
    # feature in each cluster. The color represents the average expression level
    DotPlot(pbmc3k.final, features = features) + RotatedAxis()
    ```

    ```{r}
    # SplitDotPlotGG has been replaced with the `split.by` parameter for DotPlot
    DotPlot(pbmc3k.final, features = features, split.by = "groups") + RotatedAxis()
    ```

5.  HeatMap

    ```{r}
    # Single cell heatmap of feature expression
    DoHeatmap(subset(pbmc3k.final, downsample = 100), features = features, size = 3)
    ```

    ```{r}
    # DoHeatmap now shows a grouping bar, splitting the heatmap into groups or clusters. This can
    # be changed with the `group.by` parameter
    DoHeatmap(pbmc3k.final, features = VariableFeatures(pbmc3k.final)[1:100], cells = 1:500, size = 4,
        angle = 90) + NoLegend()
    ```

6.  Dimplot

```{r}
# DimPlot replaces TSNEPlot, PCAPlot, etc. In addition, it will plot either 'umap', 'tsne', or
# 'pca' by default, in that order
DimPlot(pbmc3k.final)
```
